OpenEphys.Onix/OpenEphys.Onix/Headstage64OpticalStimulatorTrigger.cs

```
@@ -0,0 +1,225 @@
+        [Range(0.0, 1000.0)]
+        [Precision(3, 1)]
+        public double PulsesPerSecond
+        {
+            get => pulsesPerSecond.Value;
+            set => pulsesPerSecond.OnNext(value);
+        }
+
+        [Description("Number of pulses to deliver in a burst.")]
+        [Editor(DesignTypes.NumericUpDownEditor, DesignTypes.UITypeEditor)]
+        [Range(0, int.MaxValue)]
+        [Precision(0, 1)]
+        public uint PulsesPerBurst
+        {
+            get => pulsesPerBurst.Value;
+            set => pulsesPerBurst.OnNext(value);
+        }
+
+        [Description("Inter-burst interval (msec).")]
+        [Editor(DesignTypes.SliderEditor, DesignTypes.UITypeEditor)]
+        [Editor(DesignTypes.NumericUpDownEditor, DesignTypes.UITypeEditor)]
+        [Range(0.0, 10000.0)]
+        [Precision(3, 1)]
+        public double InterBurstInterval
+        {
+            get => interBurstInterval.Value;
+            set => interBurstInterval.OnNext(value);
+        }
+
+        [Description("Number of bursts to deliver in a train.")]
+        [Editor(DesignTypes.NumericUpDownEditor, DesignTypes.UITypeEditor)]
+        [Range(0, int.MaxValue)]
+        [Precision(0, 1)]
+        public uint BurstsPerTrain
+        {
+            get => burstsPerTrain.Value;
+            set => burstsPerTrain.OnNext(value);
+        }
+
+        [Description("Delay between issue of trigger and start of train (msec).")]
+        [Editor(DesignTypes.NumericUpDownEditor, DesignTypes.UITypeEditor)]
+        [Range(0.0, 1000.0)]
+        [Precision(3, 1)]
+        public double Delay
+        {
+            get => delay.Value;
+            set => delay.OnNext(value);
+        }
+
+        // TODO: Should this be checked before TRIGGER is written to below and an error thrown if
+        // DC current is too high? Or, should settings be forced too keep DC current under some value?
+        [Description("Direct current required during burst (mA). Should be less than 50 mA.")]
```

Right now, this is a read-only report of the DC current draw of the circuit when its operating. Ideally we either want to throw an error when its too high and the user tries to trigger or to force the parameters into bounds. Not sure what is the best option.

---

My main worry is that both approaches can leave the person baffled: one because the workflow crashes from out of nowhere; the other because the change didn't have the expected effect.

This kind of validation is a good example of the advantages of having an external `Configuration` object that the sink node receives. All validation happens at the moment of constructing the configuration object, rather than crashing the sink. You can also have a graphical GUI that validates your entire stim config when you click submit but before actually trying to send to the node.

The only thing this would preclude is the possibility of dynamically changing only specific individual properties and immediately writing to registers, but this is something I am still not sure is ever really required, especially if this kind of validation is warranted.

---

- GUI: Yes, this planned.
- Config Object: real-time manipulation by upstream workflow logic then because a huge pain because inner properties cannot be exposed as reactive elements. What you state about changing individual properties is probably not correct: e.g. one of them is the frequency of stimulation. This is a pretty obvious target for real-time control.

For now, I'm just going to leave as is and GUI will use to validate initial parameters. Real-time changes can result in large current draws and this might affect function I guess.